include/ginkgo/core/base/batch_multi_vector.hpp

```diff
@@ -0,0 +1,800 @@
+class BatchMultiVector
+    : public EnablePolymorphicObject<BatchMultiVector<ValueType>>,
+      public EnablePolymorphicAssignment<BatchMultiVector<ValueType>>,
+      public EnableCreateMethod<BatchMultiVector<ValueType>>,
+      public ConvertibleTo<BatchMultiVector<next_precision<ValueType>>>,
+      public BatchReadableFromMatrixData<ValueType, int32>,
+      public BatchReadableFromMatrixData<ValueType, int64>,
+      public BatchWritableToMatrixData<ValueType, int32>,
+      public BatchWritableToMatrixData<ValueType, int64> {
+    friend class EnableCreateMethod<BatchMultiVector>;
+    friend class EnablePolymorphicObject<BatchMultiVector>;
+    friend class BatchMultiVector<to_complex<ValueType>>;
+
+public:
+    using BatchReadableFromMatrixData<ValueType, int32>::read;
+    using BatchReadableFromMatrixData<ValueType, int64>::read;
+    using EnablePolymorphicAssignment<BatchMultiVector>::convert_to;
+    using EnablePolymorphicAssignment<BatchMultiVector>::move_to;
+    using ConvertibleTo<
+        BatchMultiVector<next_precision<ValueType>>>::convert_to;
+    using ConvertibleTo<BatchMultiVector<next_precision<ValueType>>>::move_to;
+
+    using value_type = ValueType;
+    using index_type = int32;
+    using unbatch_type = matrix::Dense<ValueType>;
+    using mat_data = gko::matrix_data<ValueType, int64>;
+    using mat_data32 = gko::matrix_data<ValueType, int32>;
+    using absolute_type = remove_complex<BatchMultiVector<ValueType>>;
+    using complex_type = to_complex<BatchMultiVector<ValueType>>;
+
+    using row_major_range = gko::range<gko::accessor::row_major<ValueType, 2>>;
+
+    /**
+     * Creates a BatchMultiVector matrix with the configuration of another
+     * BatchMultiVector matrix.
+     *
+     * @param other  The other matrix whose configuration needs to copied.
+     */
+    static std::unique_ptr<BatchMultiVector> create_with_config_of(
+        ptr_param<const BatchMultiVector> other)
+    {
+        // De-referencing `other` before calling the functions (instead of
+        // using operator `->`) is currently required to be compatible with
+        // CUDA 10.1.
+        // Otherwise, it results in a compile error.
+        return (*other).create_with_same_config();
+    }
+
+    friend class BatchMultiVector<next_precision<ValueType>>;
+
+    void convert_to(
+        BatchMultiVector<next_precision<ValueType>>* result) const override;
+
+    void move_to(BatchMultiVector<next_precision<ValueType>>* result) override;
+
+    void read(const std::vector<mat_data>& data) override;
+
+    void read(const std::vector<mat_data32>& data) override;
+
+    void write(std::vector<mat_data>& data) const override;
+
+    void write(std::vector<mat_data32>& data) const override;
+
+    /**
+     * Unbatches the batched dense and creates a std::vector of Dense matrices
+     *
+     * @return  a std::vector containing the Dense matrices.
+     */
+    std::vector<std::unique_ptr<unbatch_type>> unbatch() const
```

Comment:
In most cases, we have a huge number of small problems, instantiating a polymorphic type for each of them is quite inefficient. This seems mostly useful for testing, does it have other (user-facing) applications? I would suggest to have functions we use for testing kept inside test-specific headers first, and consider other more efficient ways of exposing this if necessary.

Reply:
Yes, unfortunately, applications have asked for this. Though it is definitely more of a utility function rather than a performance related function.

Reply:
Did they ask specifically for this interface, or for a way to extract single systems from a batch system? Because due to the data layout, I think it would be easier if we allowed users to access a view to a single vector of the batch vector, instead of copying them all out. Something like `std::unique_ptr<Dense> batch_vector->create_single_view(batch_id)`. The same is possible for batch matrices, and incurs essentially no performance overhead.